### PR TITLE
chore: update TypeScript to 5.4.2

### LINKED ISSUE #4275: @ngrx/effects/src/effect_creator.d.ts:12:43 - error TS2313: Type parameter 'OT' has a circular constraint.
### Which @ngrx/* package(s) are the source of the bug?

effects

### Minimal reproduction of the bug/regression with instructions

When trying to compile code in an Nx monorepo I get the following Error:
```12 }, DT extends DispatchType<C>, OT extends ObservableType<DT, OT>, R extends EffectResult<OT>>(source: () => R & ConditionallyDisallowActionCreator<DT, R>, config?: C): R & CreateEffectMetadata;```

### Expected behavior

It should not throw a TypeScript Error.
I have tried adding to the the following to the base tsconfig
```
  "compilerOptions": {
    "forceConsistentCasingInFileNames": true,
    "strict": false,
    "noImplicitReturns": false,
    "noFallthroughCasesInSwitch": true,
    "paths": {
      "@ngrx/effects": ["node_modules/@ngrx/effects"],
      "@ngrx/effects/*": ["node_modules/@ngrx/effects/*"]
    },
  },
  ```
 There are, however, dozens of tsconfigs, as this is a monorepo. They mostly inherit form a base config, but not all.
I may be able to add it to one of those configs and have it work. It would be preferable that the code did not trigger TS type checking.
  

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

Node   : 18.19.1
OS     : darwin-arm64
yarn   : 1.22.21

nx (global)        : 18.0.7
nx                 : 18.0.7
@nx/js             : 18.1.0
@nx/jest           : 18.1.0
@nx/linter         : 18.1.0
@nx/eslint         : 18.1.0
@nx/workspace      : 18.1.0
@nx/angular        : 18.1.0
@nx/cypress        : 18.1.0
@nx/devkit         : 18.1.0
@nx/eslint-plugin  : 18.1.0
@nrwl/tao          : 18.0.7
@nx/web            : 18.1.0
@nx/webpack        : 18.1.0
typescript         : 5.4.2
---------------------------------------
Community plugins:
@ng-bootstrap/ng-bootstrap : 16.0.0
@ngrx/component-store      : 17.1.1
@ngrx/effects              : 17.1.1
@ngrx/entity               : 17.1.1
@ngrx/router-store         : 17.1.1
@ngrx/schematics           : 17.1.1
@ngrx/store                : 17.1.1
@ngrx/store-devtools       : 17.1.1
ng-mocks                   : 12.5.1
---------------------------------------
The following packages should match the installed version of nx
  - @nx/js@18.1.0
  - @nrwl/js@18.1.0
  - @nx/jest@18.1.0
  - @nrwl/jest@18.1.0
  - @nx/linter@18.1.0
  - @nx/eslint@18.1.0
  - @nx/workspace@18.1.0
  - @nrwl/workspace@18.1.0
  - @nx/angular@18.1.0
  - @nrwl/angular@18.1.0
  - @nx/cypress@18.1.0
  - @nrwl/cypress@18.1.0
  - @nx/devkit@18.1.0
  - @nrwl/devkit@18.1.0
  - @nx/eslint-plugin@18.1.0
  - @nrwl/eslint-plugin-nx@18.1.0
  - @nx/web@18.1.0
  - @nrwl/web@18.1.0
  - @nx/webpack@18.1.0
  - @nrwl/webpack@18.1.0

### Other information

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### LINKED ISSUE #4279: `patchState` TypeScript v5.4 compilation error
### Which @ngrx/* package(s) are the source of the bug?

signals

### Minimal reproduction of the bug/regression with instructions

https://stackblitz.com/edit/stackblitz-starters-weemkm?file=src%2Fmain.ts

If you chain multiple updaters in `patchState`, with one being an updater from the `@ngrx/signals/entities` package, you now get a compilation error when upgrading to TypeScript v5.4.3:

```ts
type CoreState = { loading: boolean };
type Entity = { id: string };

const Store = signalStore(
  withState<CoreState>({ loading: true }),
  withEntities<Entity>(),
  withMethods((store) => {
    const setLoading = () => {
      const newState: CoreState = { loading: true };
      patchState(store, newState, removeAllEntities());
    };

    return {};
  })
);
```

```
✘ [ERROR] TS2345: Argument of type 'CoreState' is not assignable to parameter of type 'PartialStateUpdater<EntityState<any>> | Partial<EntityState<any>>'. [plugin angular-compiler]

    src/main.ts:20:24:
      20 │       patchState(store, newState, removeAllEntities());
```

This previously worked fine in TypeScript v5.3.

### Expected behavior

To compile (and work) fine.

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

NgRx Signals: v17.1.1
Angular: v17.3.0
TypeScript: v5.4.3

### Other information

_No response_

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by timdeschryver**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[x] Other... Please describe: Update dependencies
```

## What is the current behavior?

Closes #4275
Closes #4279

## Does this PR introduce a breaking change?

```
[x] Yes
[] No
```

Because `NoInfer` is used to resolve #4279 that this will result in a breaking change for users that are on TypeScript 5.3.x

### CODE REVIEW

File: modules/signals/src/patch-state.ts [hunk 1]
```diff
@@ -6,7 +6,9 @@ export type PartialStateUpdater<State extends object> = (
 
 export function patchState<State extends object>(
   stateSignal: StateSignal<State>,
-  ...updaters: Array<Partial<State & {}> | PartialStateUpdater<State & {}>>
+  ...updaters: NoInfer<
```
Comment (timdeschryver) [inline modules/signals/src/patch-state.ts:9]: NoInfer to resolve #4279
We already have a test case for this at https://github.com/ngrx/platform/blob/bc49af13bcef576f499dbadd4466ea65469fbac1/modules/signals/entities/spec/with-entities.spec.ts#L74, which fails without this fix.

This is a breaking change for the users that aren't on TypeScript 5.4.x

File: modules/effects/src/effect_creator.ts [hunk 2]
```diff
@@ -23,8 +23,9 @@ type ConditionallyDisallowActionCreator<DT, Result> = DT extends false
 export function createEffect<
   C extends EffectConfig & { functional?: false },
   DT extends DispatchType<C>,
-  OT extends ObservableType<DT, OT>,
-  R extends EffectResult<OT>
+  OTP,
+  R extends EffectResult<OT>,
+  OT extends ObservableType<DT, OTP>
```
Comment (timdeschryver) [inline modules/effects/src/effect_creator.ts:28]: Resolves #4279
Introduce a new type to resolve work around the circular generic type `OT`

File: modules/store/src/action_group_creator.ts [hunk 3]
```diff
@@ -138,17 +138,16 @@ export function createActionGroup<
...
-  return Object.keys(events).reduce(
-    (actionGroup, eventName) => ({
+  return Object.entries(events).reduce((actionGroup, [eventName, event]) => {
+    return {
       ...actionGroup,
+      // TODO: fix the any type
       [toActionName(eventName)]: createAction(
         toActionType(source, eventName),
-        events[eventName]
+        event as any
```
Comment (timdeschryver) [inline modules/store/src/action_group_creator.ts:147]: This is a quick-fix that we should come back at.
I wasn't able to fix without touching more types, help is appreciated here 😅
The problem is passing the correct event type to `createAction`.

File: package.json [hunk 4]
```diff
@@ -92,10 +92,10 @@
     "@angular-devkit/build-angular": "17.3.0",
     "@angular-devkit/core": "17.3.0",
     "@angular-devkit/schematics": "17.3.0",
-    "@angular-eslint/builder": "16.1.0",
+    "@angular-eslint/builder": "17.0.1",
```
Comment (timdeschryver) [inline package.json:95]: I forgot to update these in #4277

File: modules/effects/src/effect_creator.ts (hunk 2 above)
Comment (jits) [reply modules/effects/src/effect_creator.ts:28]: [nit] You likely meant #4275 😅

**Review by Rambou — APPROVED** [review]
The changes seem to be fine. Good work!

**Comment by Rambou** [comment]
Hi guys, can you take a look a this PR @markostanimirovic @brandonroberts ? Thanks in advance!

**Comment by brandonroberts** [comment]
@Rambou thanks. As this is a breaking change, support for TypeScript 5.4 will have to wait until NgRx 18 unless there's a workaround

**Comment by brandonroberts** [comment]
@davesmits yes, that's preferred. 

We'll discuss options. More specifically, `@ngrx/signals` is still in developer preview, so it allows us to make a breaking change in between major versions, and this seems like a reasonable way to do so without breaking the other packages and supporting Angular v17.3.

**Comment by timdeschryver** [comment]
Closed in favor of separate PRs without breaking changes.

**Comment by brandonroberts** [comment]
The fixes for TS 5.4.x have been released with NgRx 17.2.0. Thanks!

**Comment by Rambou** [comment]
The fix caused another regression. `Cannot find module '@ngrx/store/src/models'` @brandonroberts @timdeschryver 

```✖ Compiling with Angular sources in Ivy partial compilation mode.
../../../../node_modules/@ngrx/effects/src/effects_actions.d.ts:2:110 - error TS2307: Cannot find module '@ngrx/store/src/models' or its corresponding type declarations.

2 export declare const rootEffectsInit: import("@ngrx/store").ActionCreator<"@ngrx/effects/init", () => import("@ngrx/store/src/models").TypedAction<"@ngrx/effects/init">>;
       ````

**Comment by Roman991** [comment]
> The fix caused a regression. `Cannot find module '@ngrx/store/src/models'` @brandonroberts @timdeschryver
> 
> ```
> ../../../../node_modules/@ngrx/effects/src/effects_actions.d.ts:2:110 - error TS2307: Cannot find module '@ngrx/store/src/models' or its corresponding type declarations.
> 
> 2 export declare const rootEffectsInit: import("@ngrx/store").ActionCreator<"@ngrx/effects/init", () => import("@ngrx/store/src/models").TypedAction<"@ngrx/effects/init">>;
>        ````
> ```

same here after updating the "moduleResolution" in `tsconfig.json`
 
```  json
    "moduleResolution": "Bundler", // from  "node"
```

**Pull request closed without merging**